File: rust/src/generator.rs
use crate::types::{Order, Side};

pub struct Xorshift64 {
    state: u64,
}

impl Xorshift64 {
    pub fn new(seed: u64) -> Self {
        let s = if seed == 0 { 0xDEADBEEFCAFE1234 } else { seed };
        Self { state: s }
    }

    #[inline(always)]
    pub fn next(&mut self) -> u64 {
        let mut v = self.state;
        v ^= v << 13;
        v ^= v >> 7;
        v ^= v << 17;
        self.state = v;
        v
    }
}

pub fn generate_orders(n: usize, seed: u64) -> Vec<Order> {
    let mut rng = Xorshift64::new(seed);
    let mut orders = Vec::with_capacity(n);

    for i in 0..n {
        let r1 = rng.next();
        let r2 = rng.next();
        let r3 = rng.next();

        let side = if (r1 & 1) == 0 { Side::Buy } else { Side::Sell };
        let price_offset = (r2 % 200) as i64 - 100;
        let price = (10000i64 + price_offset) as u64;
        let qty = 1 + (r3 % 100);

        orders.push(Order {
            id: (i + 1) as u64,
            price,
            quantity: qty,
            side,
        });
    }

    orders
}

File: go/generator.go
package main

// Xorshift64 — побитово тот же, что golang/generator.go и rust/src/generator.rs.
type Xorshift64 struct{ state uint64 }

func NewXorshift64(seed uint64) *Xorshift64 {
	if seed == 0 {
		seed = 0xDEADBEEFCAFE1234
	}
	return &Xorshift64{state: seed}
}

func (x *Xorshift64) Next() uint64 {
	v := x.state
	v ^= v << 13
	v ^= v >> 7
	v ^= v << 17
	x.state = v
	return v
}

func GenerateOrders(n int, seed uint64) []OrderInput {
	rng := NewXorshift64(seed)
	orders := make([]OrderInput, n)
	for i := 0; i < n; i++ {
		r1, r2, r3 := rng.Next(), rng.Next(), rng.Next()
		orders[i] = OrderInput{
			ID:       uint64(i + 1),
			Price:    uint64(10000 + int64(r2%200) - 100),
			Quantity: 1 + (r3 % 100),
			Side:     Side(r1 & 1),
		}
	}
	return orders
}
